// 10tcl/lib/model/user.js
module.exports = {
    name: 'user',
    label: 'Usuários',
    format: '{{name}} {{lastname}}',
    onlyFor: ['admin', 'landlord'],
    routeTo10tcl: true,
    fields: [
        { name: 'usr', label: 'Usuário', checks: ['hasValue', 'unique', 'unchanged'] },
        { name: 'pwd', label: 'Senha', type: 'password', checks: 'hasValue' },
        { name: 'name', label: 'Nome', checks: 'hasValue' },
        { name: 'lastName', label: 'Sobrenome', checks: 'hasValue' },
        { name: 'email', label: 'e-mail', type: 'email', checks: 'hasValue' },
        { name: 'role', label: 'Perfil', onlyFor: 'admin' }
    ],

    checkUser: function( tenant, credential, afterCheck ){
        
        function afterFind( err, data ){
            if ( err ) {
                afterCheck( 500, err )
            } else {
                if ( data.length > 0 ) {
                    afterCheck( 200, JSON.parse( JSON.stringify(data) )[0] )
                } else {
                    afterCheck( 401 )
                }
            }
        }
        
        this.collection( tenant ).find( credential ).toArray( afterFind )
    },

    mock: []
}
